Libraries/LibCore/Account.cpp

```diff
@@ -0,0 +1,181 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Base64.h>
+#include <AK/Random.h>
+#include <LibCore/Account.h>
+#include <LibCore/File.h>
+#include <grp.h>
+#include <pwd.h>
+#include <string.h>
+#include <unistd.h>
+
+namespace Core {
+
+static String get_salt()
+{
+    char random_data[12];
+    AK::fill_with_random(random_data, sizeof(random_data));
+
+    StringBuilder builder;
+    builder.append("$5$");
+    builder.append(encode_base64(ReadonlyBytes(random_data, sizeof(random_data))));
+
+    return builder.build();
+}
+
+static Vector<gid_t> get_gids(const StringView& username)
+{
+    Vector<gid_t> extra_gids;
+    for (auto* group = getgrent(); group; group = getgrent()) {
+        for (size_t i = 0; group->gr_mem[i]; ++i) {
+            if (username == group->gr_mem[i])
+                extra_gids.append(group->gr_gid);
+        }
+    }
+    endgrent();
+    return extra_gids;
+}
+
+Result<Account, String> Account::get(const StringView& username)
+{
+    String username_with_terminator(username);
+
+    struct passwd* pwd = nullptr;
+    pwd = getpwnam(username_with_terminator.characters());
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+Result<Account, String> Account::get(uid_t uid)
+{
+    struct passwd* pwd = nullptr;
+    pwd = getpwuid(uid);
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+bool Account::authenticate(const StringView& password) const
+{
+    if (m_passwd == "")
+        return true;
+
+    String password_with_terminator(password);
+
+    // FIXME: Use crypt_r if it can be built in lagom.
+    char* hash = crypt(password_with_terminator.characters(), m_passwd.characters());
+    return hash != nullptr && strcmp(hash, m_passwd.characters()) == 0;
+}
+
+Result<uid_t, int> Account::login() const
+{
+    if (setgroups(m_extra_gids.size(), m_extra_gids.data()) < 0)
+        return errno;
```

Comment:
I don't trust type-based dispatch to work too well here (what if `uid_t` is an `int`? or if `errno` is not an `int`, but say an `unsigned int` or an `int32_t`?)

Is there an explicit way to initialize one of the `Result` variants?

Reply:
Honestly, there is no reason to return the `uid_t` in the first place.

Is there an idiomatic way to return just an error in serenity?

Reply:
In the kernel, that'd be `KResult`. In userspace, I don't think we have anything quite like that. We could just do `bool` + `errno` though.